### DISCUSSION TITLE
ssh: firejailing ssh connection

### DISCUSSION THREAD

**Discussion by Rosika2** [netblue30/firejail#4773] [Q&A] [answered]
Hi all,

I´ve got a question just for better understanding of how things work - so no real problem there.

My main system is Linux/Lubuntu 20.04.3.
On it I´m running KVM/qemu/virt-manager; and here BodhiLinux is installed as a VM.

Often enough I don´t need a graphical interface and so I get the VM running thus: 

`firejail ssh rosika2@192.168.122.76  # rosika2 being the VM`

This works quite well (and firejail provides ssh profiles, which is very nice).

Now I´ve found out that I can also use X forwarding by invoking the command 

`firejail ssh -X rosika2@192.168.122.76`

on my main system because on Bodhi (the VM) the "X11Forwarding" option is enabled in the **/etc/ssh/sshd_config** file.

And it works just fine.

When firing up "**thunar**" file manager per command in the terminal ("firejail ssh -X rosika2@192.168.122.76" gives me that terminal) I get the thunar file manager from my Bodhi VM on my host (Lubuntu).
That´s really great.

What I want to know - and just to make sure - is the following:

Is thunar (started in the described way) still sandboxed?

My understanding is: it should be, as the original parent-command ("firejail ssh -X rosika2@192.168.122.76") is firejailed anyway.

Plus:

On Lubuntu:

```
rosika@rosika-10159 ~> ps aux | grep thunar
rosika     27331  0.0  0.0   9412   664 pts/4    S+   17:00   0:00 grep --color=auto thunar
```

suggests **no thunar instance is running on my host**.

I´m just asking because "firejail --tree" just gives the following three items as output, _regardless of whether thunar is running in the VM or not_:

```
26325:rosika::firejail ssh -X rosika2@192.168.122.76 
  26326:rosika::firejail ssh -X rosika2@192.168.122.76 
    26329:rosika::ssh -X rosika2@192.168.122.76 
```
    
Is it just supposed to be this way?

Thanks so much for your help in advance.

Many greetings.
Rosika

**Comment by ghost** [comment, marked as answer]
> What I want to know - and just to make sure - is the following:
Is thunar (started in the described way) still sandboxed?

Firejail implements `profile inheritance` in cases where a specific application isn't explicitly blacklisted (like blacklist ${PATH}/thunar or thunar not being in a private-bin). So yes, in your described way thunar inherits the ssh sandbox profile, instead of its own thunar.profile --> Thunar.profile --> file-manager-common.profile. This ssh sandbox isn't specifically designed for thunar though, meaning it could be hardened. Whether or not that's worthwhile will always depend on your specific usage of thunar under those conditions and be a matter of degree. Firejail's file managers sandboxes aren't very strong to begin with, as they are conceived rather permissive for functionality's sake. That's also why they are not enabled in firecfg by default IMO.

I don't think this is explained somewhere in detail (like in the README.md or the man page) or in a wiki page (like under the FAQ section). Apparently the --tree option isn't very helpful either to detect sandbox inheritance. Not sure whether this could be made more discoverable for users in some way...

**Reply by Rosika2** [reply]
@glitsj16:

Hello and thank you so much for your reply,

> So yes, in your described way thunar inherits the ssh sandbox profile, instead of its own thunar.profile --> Thunar.profile --> file-manager-common.profile.

I see. 
Well, that´s good to know.

I suppose special hardening shouldn´t be neccessary. Still I´m glad to learn _thunar_ runs sandboxed anyway.
Besides running _thunar_ was just an example. The text editor **Kate** which is also installed on the VM would be another example.

I was just wondering about the matter as a whole as `ps aux | grep thunar` (issued on my **host**) didn´t suggest that thunar was running on the host.
From this I inferred that the running thunar instance was (and of course should be) triggered by the VM.
So it´s part of the VM and not of the host.

As a VM provides some kind of sandboxing in itself PLUS the fact that I issued the _ssh_-command **firejailed** 
I think that should provide a high enough degree of sanboxing as it is.

I hope I´m not completely wrong there.

Many thanks for your help and many greetings.

Rosika

**Comment by zithro** [comment]
I know this is solved but I'm not sure the answer is right ? Local firejail cannot protect remote applications.

When running `user@host $ firejail ssh -X user@remotehost APP`, you're sandboxing the application that runs on the **host**, in this case `ssh`.
When APP is run from such remote shell, APP runs on the **remote host**, so it's not sandboxed at all by default.
That's why, in this case, `firejail -tree` will never show remote applications. Try running `htop` in the remote host, APP will show up as a descendant of `sshd` (the ssh server).

I think the correct way is to run `user@host $ firejail ssh -X user@remotehost firejail APP`.
This way, ssh is sandboxed in the host, and APP is sandboxed on the remote-host.
Of course, firejail needs to be also installed in the remote host.

**Comment by Rosika2** [comment]
@zithro:

Hi and thanks so much for the update on the matter.  :-)

> When running `user@host $ firejail ssh -X user@remotehost APP,` you're sandboxing the application that runs on the host, in this case `ssh`.

I see. Yes, that makes perfect sense.

>  I think the correct way is to run `user@host $ firejail ssh -X user@remotehost firejail APP`

That sounds logical indeed. So I tried it out but there seems to be a problem with displaying the APP on the host.
These were my steps:

`firejail ssh -X rosika2@192.168.122.174 # debian VM already running`

then trying to sandbox **thunar** actually fails:

```
rosika2@debian ~> firejail thunar
Reading profile /etc/firejail/thunar.profile
Reading profile /etc/firejail/Thunar.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-passwdmgr.inc
Warning: networking feature is disabled in Firejail configuration file
Parent pid 1097, child pid 1098
Blacklist violations are logged to syslog
Child process initialized in 70.01 ms
Unable to init server: Verbindung ist gescheitert: Verbindungsaufbau abgelehnt

(thunar:3): Gtk-WARNING **: 14:46:58.579: cannot open display: localhost:10.0

Parent is shutting down, bye...

```
But 
`firejail --noprofile thunar` works of course and even
`firejail --profile=/etc/firejail/default.profile thunar` works.

So I guess something has to be changed in `thunar.profile` for it to work the way you described.

BTW: same behaviour with `mousepad`.

Thanks a lot and many greetings.
Rosika

**Reply by kmk3** [reply]
(Offtopic)

@Rosika2 [on Mar 15](https://github.com/netblue30/firejail/discussions/4773#discussioncomment-2365362):

> ```
> rosika2@debian ~> firejail thunar
> [...]
> Unable to init server: Verbindung ist gescheitert: Verbindungsaufbau abgelehnt
> ```

Please always prefix `firejail` with `LC_ALL=C` when posting code.  Example:
`LC_ALL=C firejail thunar`.  That prevents localization, which thus makes the
errors more understandable/searchable in general.

**Reply by Rosika2** [reply]
@kmk3:

Hi Kelvin,   :-)

> Please always prefix firejail with LC_ALL=C

Thanks for bringing that to my attention. Sorry I didn´t think of that earlier.

As I´m using `fish` as my default shell I even have to do it this way:
`env LC_ALL=C firejail thunar`

O.K. then. Here´s the output:

```
env LC_ALL=C firejail thunar
Reading profile /etc/firejail/thunar.profile
Reading profile /etc/firejail/Thunar.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-passwdmgr.inc
Warning: networking feature is disabled in Firejail configuration file
Parent pid 903, child pid 904
Blacklist violations are logged to syslog
Child process initialized in 68.12 ms
Unable to init server: Could not connect: Connection refused

(thunar:3): Gtk-WARNING **: 14:05:46.856: cannot open display: localhost:10.0

Parent is shutting down, bye...

```

Hope it helps.

Thanks again and many greetings.
Rosika  :-)

**Comment by zithro** [comment]
I think the solution lies here : `cannot open display: localhost:10.0`. This means thunar is fordbidden access to localhost:6010.

Are you sure you're not using a custom profile for thunar restricting inet access ?
In "/etc/firejail/file-manager-common.profile", I can read `protocol unix,inet,inet6,netlink`, so it should be allowed by default.

On the other hand, mousepad is not allowed by default to access inet sockets.
From "/etc/firejail/mousepad.profile" : `protocol unix`. Change it to `protocol unix,inet`, then it works.

I only specified `inet` because it seems you're using IPv4 (rosika2@192.168.122.174), if also using IPv6, you should also add `inet6`.

**Reply by Rosika2** [reply]
Hi @zithro,

thanks for your help.

> Are you sure you're not using a custom profile for thunar restricting inet access ?

Well, I´m pretty sure I didn´t change the default settings.
Anyhow I looked it up and `thunar` reads the `Thunar.profile` (it´s redirected from `thunar.profile`.
It looks like this:

```
# Firejail profile for Thunar
# Description: File Manager for Xfce
# This file is overwritten after every install/update
# Persistent local customizations
include Thunar.local
# Persistent global definitions
include globals.local

noblacklist ${HOME}/.local/share/Trash
noblacklist ${HOME}/.config/Thunar
noblacklist ${HOME}/.config/xfce4/xfconf/xfce-perchannel-xml/thunar.xml

include disable-common.inc
include disable-devel.inc
include disable-interpreters.inc
include disable-passwdmgr.inc
# include disable-programs.inc

caps.drop all
netfilter
no3d
nodvd
nogroups
nonewprivs
noroot
nosound
notv
novideo
protocol unix
seccomp
shell none
tracelog

```
I was looking around but couldn´t find out where _**/etc/firejail/file-manager-common.profile**_  would kick in ...
Perhaps  we should do the same here as you suggested for `mousepad`, i.e. 

> protocol unix. Change it to protocol unix,inet

???

As far as `mousepad` is concerned: great success!!! Splendid piece of advice from you. Now it works.

Thanks very much.  :-)

Many greetings from Rosika

**Reply by Rosika2** [reply]
UPDATE:

I´ve just tried it and did the same for `Thunar.profile`.
I changed the respective entry to `protocol unix,inet`.
**It works as desired!!!**  :-)

So it seems the same soultion (like the one for `mousepad` applies here.)
I hope I did the right thing.

Thanks so much for your help.

Many greetings from Rosika   :-)

**Comment by zithro** [comment]
I just wanna point out a few remarks ^^

Don't ever change files in /etc/firejail, you'll lose time for no benefit (EDIT: it's written in the file: "# This file is overwritten after every install/update)".
The next upgrade will wipe your customizations (well not really as dpkg-old is your friend, but you'll have to re-merge them).
The proper way to do it is to create a "thunar|mousepad.local" in your home, and only include the stuff you're interested in, in your case `protocol unix,inet` (or even only `protocol inet`, but not tested).

Also, I see you're not using the last version of the package "firejail-profiles".
I can tell because in my "Thunar.profile", I have "include file-manager-common.profile".

So I don't know if you installed from stable or backports, but when installing from backports, it will install the **last version of firejail, but an older version of firejail-profiles**. This will mess up a few things.
You'll have to install afterwards firejail-profiles manually from backports for the thing to be consistent.
I've warned the firejail backport maintainer on Debian so it may be resolved in the next packages versions, but don't know how he will handle that because it's a bit complicated. But he's very responsive though.

So for anyone else coming here and using Debian backports, if you do `apt install firejail/bullseye-backports`, immediately do `apt install firejail-profiles/bullseye-backports` afterwards for a nice experience.

**Reply by Rosika2** [reply]
Hi @zithro:

Thanks a lot for your additional remarks. They´re highly appreciated.  :-)

> Don't ever change files in /etc/firejail, you'll lose time for no benefit (EDIT: it's written in the file: "# This file is overwritten after every install/update)".

O.K. That´s clear to me. In fact I didn´t change anything in `/etc/firejail` but copied `Thunar.profile` thus:
`cp /etc/firejail/Thunar.profile ~/.config/firejail/`
I always  do such copying for experiments. I just forgot to mention it in my previous post. Sorry.  :-(
See:

```
rosika2@debian ~> firejail thunar
Reading profile /etc/firejail/thunar.profile
Reading profile /home/rosika2/.config/firejail/Thunar.profile  # **here my modified profile is read**
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-passwdmgr.inc
Warning: networking feature is disabled in Firejail configuration file
Parent pid 1143, child pid 1144
Blacklist violations are logged to syslog
Child process initialized in 77.53 ms

```

> Also, I see you're not using the last version of the package "firejail-profiles".
> I can tell because in my "Thunar.profile", I have "include file-manager-common.profile".

Ah, I see. No wonder then I couldn´t find the `file-manager-common.profile`.

> So I don't know if you installed from stable or backports

Uh, you got me there... 
It´s been a while since I installed debian in a VM. I know it´s the stable version.

```
lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 10 (buster)
Release:	10
Codename:	buster
```
Some other info I can provide is the output of
`grep -r -v -e ^$ -e ^# /etc/apt/sources.list /etc/apt/sources.list.d/`


```
/etc/apt/sources.list:deb http://deb.debian.org/debian/ buster main
/etc/apt/sources.list:deb-src http://deb.debian.org/debian/ buster main
/etc/apt/sources.list:deb http://security.debian.org/debian-security buster/updates main
/etc/apt/sources.list:deb-src http://security.debian.org/debian-security buster/updates main
/etc/apt/sources.list:deb http://deb.debian.org/debian/ buster-updates main
/etc/apt/sources.list:deb-src http://deb.debian.org/debian/ buster-updates main
/etc/apt/sources.list.d/anydesk-stable.list:deb http://deb.anydesk.com/ all main
```
Is there another way of finding out the respective information?
I´ll be glad to look further...

Thanks for clearing things up.

Many greetings
Rosika  :-)

**Comment by zithro** [comment]
> Thanks a lot for your additional remarks. They´re highly appreciated. :-)
> [...]
> I always do such copying for experiments. I just forgot to mention it in my previous post. Sorry. :-(

No prob !

> It´s been a while since I installed debian in a VM. I know it´s the stable version
> Description:	Debian GNU/Linux 10 (buster)

Stable, yeah, but not the current one ! Buster is oldstable, it's time to upgrade to Debian 11 aka bullseye ; )
You don't have to reinstall, just upgrade in place.

> Some other info I can provide is the output of
> `grep -r -v -e ^$ -e ^# /etc/apt/sources.list /etc/apt/sources.list.d/`

You're not using backports then. If so, you would have had lines like `deb http://mirror/debian bullseye-backports [...]` (note though this is a backports line for bullseye, not buster).
More info about backports : https://backports.debian.org and https://backports.debian.org/Instructions

> Is there another way of finding out the respective information?

If you wanna know the currently installed firejail version, use `dpkg -l firejail` and `dpkg -l firejail-profiles`.
Packages installed through backports have "~bpo" in their names.

(EDIT: buster-backports -> bullseye-backports in apt source example line)

**Comment by Rosika2** [comment]
Hi @zithro,   :-)

thanks a lot for your continued help. That´s really great.

O.K, I knew Buster is oldstable. But my train of thought was thus:

I´m using as an LTS-version. Support for Buster would be available until 2022-07 but I may receive further updates from July, 2022 to June, 2024 by the _security and release teams_ (according to https://wiki.debian.org/LTS ).

As I´m using debian in a **VM** I thought I could wait a little longer.
But you suggestion as far as upgrading to bullseye is concerned is certainly worthwhile.

>  You don't have to reinstall, just upgrade in place.

That´s great. I have to admit though I´ve never done that before. 
When I need a new version of Lubuntu (my daily driver) I always do a fresh/clean install although an inline-upgrade would theoretically be possible there as well. It´s generally assumed a fresh/clean install is somewhat safer...
Yet for my debian VM I might try an in place-upgrade for a  change.

> You're not using backports then.
That was my understanding as well. Thanks for the confirmation.

Also thanks a lot for providing the links regarding backports. I´ll look into them.  :-)

As far as the firejail version is concerned I can now provide the respective info (thanks to your help):

```
firejail             0.9.58.2-2+deb10u2 amd64 
firejail-profiles    0.9.58.2-2+deb10u2 all 
```
So the packages seem to match.

Thanks so much for the wealth of information you provided. 
I´m always glad to learn something new.

Many greetings from Rosika  :-)

**Comment by zithro** [comment]
No problem, happy to help !
Concerning the version you use, I only made a recommendation, you do as you please ^^
I didn't know about LTS versions (thx for the link), but from what I've quickly read it's mainly about security. So you may not have the last version of packages, especially web browsers.
I use Debian inline-updates since 3 or 4 versions on many different systems without any problems : servers and workstations, bare metal and virtualized, x86_64 and arm (Pi 2/3).
You just have to carefully read the documentation which gives detailed steps and extensive details and glitches.
I've never really compared an updated and a clean system, but it's as clean as `apt` can get !

To show you why backports can be interesting, here is my `dpkg -l "firejail*"` output : 
```
firejail          0.9.68-3~bpo11+1 amd64
firejail-profiles 0.9.68-3~bpo11+1 amd64
```

Have fun !

**Reply by Rosika2** [reply]
@zithro:

Hi again, :-)
thanks for your reply.

> Concerning the version you use, I only made a recommendation

I´m very grateful for that. 
Good to know you have such great experience with inline-upgrades. I´d like to try it but it would be good to have a rough estimate of how much data - download-wise - it takes.
My monthly "allowance" (using a 4G-stick as my sole means of internet connection) is 5 GB/28 days.
At the end of my option period I usually have 1 - 1.5 GB left which I could "afford" to use for inline upgrade purposes ...

> You just have to carefully read the documentation which gives detailed steps and extensive details and glitches

That I´ll do. Thx a lot.
Thanks to your your output of `dpkg -l "firejail*"` I see that using backports may very well be beneficial.

Thank you very much for your help and good advice. It´s highly appreciated.
Many greetings from Rosika  :-)

**Reply by zithro** [reply]
> I´d like to try it but it would be good to have a rough estimate of how much data - download-wise - it takes

You could use `apt --dry-run` to know how much data is needed. apt/apt-get always tell you how much they will need to download.
dry-run means "display what can be done but don't do it". You can read it in `man apt-get`.
But again, carefully read the update procedure first, as it involves using two steps (`apt upgrade --without-new-pkgs` then `apt full-upgrade`).
Roughly, you could : 
- update your sources.list "buster -> bullseye"
- do `apt update` then `apt --dry-run full-upgrade`
- if not ready to update, re-change your sources.list (bullseye->buster) and **immediately** do an `apt update` !

Here are examples of two of my systems, from buster to bullseye. Both systems are Debian with Xfce -and- MATE installed.
System 1
`apt upgrade --without-new-pkgs` : 374 MB (806 upgraded, 0 newly installed, 0 to remove and 608 not upgraded)
`apt full-upgrade` : 702 MB (605 upgraded, 227 newly installed, 60 to remove and 3 not upgraded)
TOTAL: 1076 MB

System 2
`apt upgrade --without-new-pkgs` : 821 MB (909 upgraded, 0 newly installed, 0 to remove and 735 not upgraded)
`apt full-upgrade` : 1083 MB (734 upgraded, 268 newly installed, 36 to remove and 1 not upgraded)
TOTAL: 1904 MB

As you can see it may vary by a lot, as it depends on what is already installed on your system.
But it's still less than downloading the full 4GB DVD, if you're only using one system.
Another solution would be to find a public hotspot to do the upgrade ^^

> My monthly "allowance" (using a 4G-stick as my sole means of internet connection) is 5 GB/28 days

Concerning your data plan, some providers allow you to use more than advertised, but the speeds will be capped once the limit is reached. But some plans charge you for the excess.
Carefully check the conditions ! ^^

**Reply by Rosika2** [reply]
Hello @zithro,  :smile: 

thank you so much for your very detailed report. That´s really great.

> apt/apt-get always tell you how much they will need to download.

I tried it but it doesn´t seem to work that way with me. :disappointed: 
As an experiment I changed the sources.list:
`sudo sed -i 's/buster/bullseye/g' /etc/apt/sources.list`
and then `sudo apt update`.
After that: `apt --dry-run full-upgrade > output.txt`, to look out for data consumption. Yet there´s nothing to be found in the file.

I get  the sections:

> - The following packages were automatically installed and are no longer required: [...]
> - The following packages will be REMOVED: [...]
> - The following NEW packages will be installed: [...]
> - The following packages have been kept back: [...]
> - The following packages will be upgraded: [...]
> - 1365 upgraded, 239 newly installed, 54 to remove and 2 not upgraded. [...]

I looked through the whole of the file but there seems to be no hint as to how much data would be downloaded.

But thanks a lot for providing the info regarding your systems. :+1: 

Referring to your system1: "TOTAL: 1076 MB":  This seems doable for me :blush:.
My Debian system is fairly basic; I haven´t installed a lot of additional  packages.
Still: it would be nice to have a pretty accurate figure with respect to data download in advance.
No idea why I cannot get the respective info... :thinking: 

Thanks very much, @zithro, for your help.

Many greetings from Rosika :smiley:

**Reply by zithro** [reply]
> I looked through the whole of the file but there seems to be no hint as to how much data would be downloaded

Sorry, my bad, I had no packages to update at the time, so assumed it would show (I got info from old log files, using the `script` trick mentionned in the Debian upgrade docs) ...
So you're right, "--dry-run" does not show this info ! Which is funky.

Two solutions then : 
- try `apt -o APT::Get::Trivial-Only=true full-upgrade` (recommended)
- run `apt full-upgrade`, but answer "No". As there will be many packages to update, apt will ask you to continue. It should work but it's not the safest choice ! Your apt options may override the question and launch the upgrade ...

End of output of `apt -o APT::Get::Trivial-Only=true full-upgrade` : 
```
[...]
1242 upgraded, 287 newly installed, 11 to remove and 0 not upgraded.
Need to get 931 MB/932 MB of archives.
After this operation, 1,213 MB of additional disk space will be used.
E: Trivial Only specified but this is not a trivial operation.
```

Also, if you wanna be on the safe side, you could disable your network after doing a `sed sources + apt update`.
So even if, for whatever reasons, apt tries to upgrade, it could not, and you can then revert the changes and redo an apt update.

**Reply by Rosika2** [reply]
Dear @zithro, :smiley: 

thank you so much for your suggestions. :smiley: 

> Sorry, my bad, I had no packages to update at the time, so assumed it would show

No problem.
I´m ever so grateful to you that you took so much time out of your day in order to help me.
It´s very much appreciated,

> So you're right, "--dry-run" does not show this info ! Which is funky.

Yes, I would´ve expected to see the info as well. As you I was pretty surprised.

`env LANG=en_US.UTF-8 sudo apt -o APT::Get::Trivial-Only=true full-upgrade`

Great; this one worked. Thanks a lot. :+1: 

I got the following output:

```
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
[...]
1365 upgraded, 239 newly installed, 54 to remove and 2 not upgraded.
Need to get 1072 MB of archives.
After this operation, 1201 MB of additional disk space will be used.
E: Trivial Only specified but this is not a trivial operation.

```
Well, I´ll be damned...
I didn´t reckon over 1 GB of data would be downloaded, bearing in mind that I have a pretty basic system with not many additional packages post-installed... :thinking: 

At any rate that´s really good to know.
Plus: I hadn´t known the command you provided before. So I´ve learnt something new today.

Thanks for your additional advice as well.
You really helped me a lot.

Many greetings from Rosika. :smiley:
